src/clj/puppetlabs/services/request_handler/request_handler_service.clj

```diff
@@ -22,12 +32,16 @@
           (log/warn (format "%s %s"
                             (i18n/trs "The ''master.allow-header-cert-info'' setting is deprecated and will be ignored in favor of the ''authorization.allow-header-cert-info'' setting because the ''jruby-puppet.use-legacy-auth-conf'' setting is ''false''.")
                             (i18n/trs "Remove the ''master.allow-header-cert-info'' setting.")))))
-      (assoc context :request-handler
-                     (request-handler-core/build-request-handler
-                      jruby-service
-                      (request-handler-core/config->request-handler-settings
-                       config)
-                      current-code-id))))
+      ;; TODO: Throw error when max-active-requests > 0, but metrics-service is nil
```

Comment:
Hmm. That is an interesting idea. I suppose the trade-off is that the helper would then duplicate a lot of logic and error handling to ensure the counter is properly incremented and decremented --- logic which is already present (and tested!) in the metrics service. Also, if we ended up using heuristics to tune the retry behavior then we'd probably want to drive that using other bits of data from the JRuby metrics service.

So, I'm not sure which is better: keeping the retry handler decoupled from the metrics or not having to re-implement something that the metrics are already tracking.